python/pyspark/sql/dataframe.py

```diff
@@ -375,6 +375,9 @@ def _truncate(self):
         return int(self.sql_ctx.getConf(
             "spark.sql.repl.eagerEval.truncate", "20"))
 
+    def __len__(self):
```

Comment:
This is roughly the same line of thinking I'd have. I don't know how expensive it is to allow `count` to be able to be called within iterators, etc.

Reply:
I'd argue for bringing this in, if you don't think we're providing people a footgun where they'd incidentally use `len()` on a dataframe often. As for making a plan around built in function support, I'm happy to be part of a `_repr_*_` campaign. I wouldn't have the background to participate in others (`__lt__`, etc.) as I wouldn't be able to weigh their maintainability, performance, and utility like I could visual elements like reprs.

Reply:
I would say `_repr_*_` was added because I saw a strong need and usecase for it. I am worried about `len(..)` we are trying to add here. For instance, what would you expect from `list(df)` and `len(df)` ..

Reply:
> what would you expect from `list(df)` and `len(df)` ..

😨

Reply:
I mean I _think_ a reasonable thing to do if someone calls `list(df)` is collect - they clearly want the dataframe as a list. If that's a good idea or not is up to the developer.

Reply:
How about `iter(df)` .. how about `df[0]` ..

Reply:
Well those are a bit harder to say, I _think_ `iter` might be reasonable (main concern is if folks tried to use `map(lambda x, df)`) but those aren't the parts of the API we're talking about right now and is starting to boarder on a broader design decision we should consider taking to the list. Given the timeline of 3 this seems like a good time to have these discussions anyways -- maybe we can look at Dask for some inspiration on how to provide a more python friendly API while still encouraging good design on the part of our users.

That being said, I think the potential confusion of `iter` or indexing into a DF shouldn't block adding other more reasonable helpers.

Reply:
Adding some more reasonable helpers can actually confuse. It might be better nothing as well. I agree with adding those at the end but I think we better have the explicit plan before adding those bit by bit.